### ISSUE TITLE
Discussion: Extending format_time() and parse_time()

### ISSUE THREAD

**Issue by mrworkman** [rsyslog/rsyslog#2698]
I started trying to put together some `strftime()`-like functionality for `format_time()` the other day, and thought there may be some discussion that needs to be done around how it should behave. I'm opening this issue as a place we can do that.

@rgerhards, I think you want to stay away from directly using `strftime()` as much as possible in implementation due to portability reasons, right?

So far, I've roughed-in support for locale-non-specific format specifiers, like %Y, %m, %d, etc... but I'm not sure which direction to go on those that *are* locale-specific. For example: %a (abbreviated name of the day), %B (full month name), etc... try to pull this from the OS (if it's portable)? Re-invent the wheel and support X number of locales? Only support one locale (e.g. en_US)?

Also, there's an issue of supporting dates that could be far in the future, or in the past (years with more than 4 digits, years that are negative). Where is a good place to draw the line?

I guess this issue should be flagged as "enhancement"?

Related to PRs #1814, and #1925.

**Comment by davidelang** [comment]
what are the portability limits of strftime()

my knee jerk reaction is that rather than re-writing strftime for all platforms, 
we may be better off using strftime and documenting that this isn't available 
for platform X yet.

there is a LOT of value in using the standard library, so unless it is extremely 
limited, we should try to use it.

**Comment by rgerhards** [comment]
Worse: it modifies the TZ variable, so all calls related to it (and potentially existing calls) would need to be mutex protected. A no-go caused by the braindead time apis...

**Comment by mrworkman** [comment]
Right, I totally forgot about how it behaves in relation to the TZ variable.  That is a huge pain...

**Comment by rgerhards** [comment]
> I'm not sure which direction to go on those that are locale-specific. For example: %a (abbreviated name of the day), %B (full month name), etc... try to pull this from the OS (if it's portable)? Re-invent the wheel and support X number of locales? Only support one locale (e.g. en_US)?

That's to complex thing. I would try to obtain this from the OS if possible. Otherwise I would simply not support it. If we have UTC and local time numerical values, that should be great for most applications.

**Comment by mrworkman** [comment]
I spent a bit of time looking into whether this locale-specific stuff can be gotten from the OS (well, GNU) without having to use `strftime()`, and it doesn't seem promising. So, maybe just the numerical values at least for the time being.

**Comment by rgerhards** [comment]
Apologies for the very late follow-up. This may no longer help the original reporter, but I am recording the current resolution for people finding this through search while working through the backlog. We do not close reports merely because they are old.

The design discussion has converged into supported, portable APIs rather than locale-dependent strftime semantics. Current RainerScript provides parse_time(), parse_time_localtz(), format_time() with documented numeric/RFC formats, and strptime_to_rfc3339() for parsing arbitrary specified input formats. Locale-name formatting was intentionally not adopted because of portability and process-global timezone concerns discussed here.

Closing this design discussion as resolved by the current API direction. Focused requests for a missing concrete format can be opened separately with input, output, timezone, and locale expectations.
